README.md
# React-web-notification [![Build Status](https://travis-ci.org/mobilusoss/react-web-notification.svg?branch=develop)](https://travis-ci.org/mobilusoss/react-web-notification) [![npm version](https://badge.fury.io/js/react-web-notification.svg)](http://badge.fury.io/js/react-web-notification) [![codebeat badge](https://codebeat.co/badges/e03e06fa-8d28-44a9-afb9-848cfcf98d91)](https://codebeat.co/projects/github-com-mobilusoss-react-web-notification-master) [![FOSSA Status](https://app.fossa.io/api/projects/git%2Bgithub.com%2Fmobilusoss%2Freact-web-notification.svg?type=shield)](https://app.fossa.io/projects/git%2Bgithub.com%2Fmobilusoss%2Freact-web-notification?ref=badge_shield) [![codecov](https://codecov.io/gh/mobilusoss/react-web-notification/branch/master/graph/badge.svg)](https://codecov.io/gh/mobilusoss/react-web-notification)

React component with [HTML5 Web Notification API](https://developer.mozilla.org/en/docs/Web/API/notification).
This component show nothing in dom element, but trigger HTML5 Web Notification API with `render` method in the life cycle of React.js.

## Demo

[View Demo](https://mobilusoss.github.io/react-web-notification/example/)

## Installation

```bash
npm install --save react-web-notification
```

## API

### `Notification`

React component which wrap web-notification.

#### Props

```javascript
Notification.propTypes = {
  ignore: bool,
  disableActiveWindow: bool,
  askAgain: bool,
  notSupported: func,
  onPermissionGranted: func,
  onPermissionDenied: func,
  onShow: func,
  onClick: func,
  onClose: func,
  onError: func,
  timeout: number,
  title: string.isRequired,
  options: object,
  swRegistration: object,
};

```

* `ignore` : if true, nothing will be happen

* `disableActiveWindow` : if true, nothing will be happen when window is active

* `askAgain` : if true, `window.Notification.requestPermission` will be called on `componentDidMount`, even if it was denied before,

* `notSupported()` : Called when [HTML5 Web Notification API](https://developer.mozilla.org/en/docs/Web/API/notification) is not supported.

* `onPermissionGranted()` : Called when permission granted.

* `onPermissionDenied()` : Called when permission denied. `Notification` will do nothing until permission granted again.

* `onShow(e, tag)` : Called when Desktop notification is shown.

* `onClick(e, tag)` : Called when Desktop notification is clicked.

* `onClose(e, tag)` : Called when Desktop notification is closed.

* `onError(e, tag)` : Called when Desktop notification happen error.

* `timeout` : milli sec to close notification automatically. Ignored if `0` or less. (Default `5000`)

* `title` : Notification title.

* `options` : Notification options. set `body`, `tag`, `icon` here.
  See also (https://developer.mozilla.org/en-US/docs/Web/API/Notification/Notification)

* `swRegistration` : ServiceWorkerRegistration. Use this prop to delegate the notification creation to a service worker.
  See also (https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerRegistration/showNotification)
  ⚠️ `onShow`, `onClick`, `onClose` and `onError` handlers won't work when notification is created by the service worker.


## Usage example

See  [example](https://github.com/mobilusoss/react-web-notification/tree/develop/example)

```bash
yarn
yarn run start:example
```

## Tests

```bash
yarn test
```

## Update dependencies

Use [npm-check-updates](https://www.npmjs.com/package/npm-check-updates)

### Known Issues

 * [Notification.sound](https://github.com/mobilusoss/react-web-notification/issues/13)
  `Notification.sound` is [not supported in any browser](https://developer.mozilla.org/en/docs/Web/API/notification/sound#Browser_compatibility).
  You can emulate it with `onShow` callback. see [example](https://github.com/mobilusoss/react-web-notification/tree/develop/example).


## License
[![FOSSA Status](https://app.fossa.io/api/projects/git%2Bgithub.com%2Fmobilusoss%2Freact-web-notification.svg?type=large)](https://app.fossa.io/projects/git%2Bgithub.com%2Fmobilusoss%2Freact-web-notification?ref=badge_large)


CHANGELOG.md
## Change Log

### Ver 0.8.0
* Modern browsers support #66

### Ver 0.7.0
* Enable to wait user interaction to close #60

### Ver 0.6.1
* Update dependencies with `ncu -u`
* Add codebeat

### Ver 0.6.0
* Update dependencies, add core-js@3
* Change url from georgeosddev to mobilusoss #48

### Ver 0.5.0

* Update `Babel` from 6.x to 7.x
* Update other dependencies

### Ver 0.4.0

* [add service worker registration prop](https://github.com/mobilusoss/react-web-notification/pull/41)

* [Add askAgain to propTypes in Readme](https://github.com/mobilusoss/react-web-notification/pull/42)

### Ver 0.3.2

* [Typo issue](https://github.com/mobilusoss/react-web-notification/issues/39)

### Ver 0.3.1

* [Change demo link to https](https://github.com/mobilusoss/react-web-notification/issues/33)

### Ver 0.3.0

* [Update React and dependencies](https://github.com/mobilusoss/react-web-notification/issues/31)

### Ver 0.2.4

* [#28 Use prop-types package instead of accessing PropTypes](https://github.com/mobilusoss/react-web-notification/issues/28), thanks @AlexNisnevich

### Ver 0.2.3

* [#19 add support for react@^15](https://github.com/mobilusoss/react-web-notification/issues/19), thanks @emirotin

### Ver 0.2.2

  * [#15 Permission being repeatedly asked for in Safari](https://github.com/mobilusoss/react-web-notification/issues/11)

### Ver 0.2.1

  * [#11 depending on library](https://github.com/mobilusoss/react-web-notification/issues/11)

### Ver 0.2.0
  * [#7 Support React-v0.14](https://github.com/mobilusoss/react-web-notification/issues/7)
  * [#8 Publish transpiled code](https://github.com/mobilusoss/react-web-notification/issues/8)

### Ver 0.1.2
  * [#5 Add option `disableActiveWindow`](https://github.com/mobilusoss/react-web-notification/issues/5)

### Ver 0.1.1
  * [#2 Compile before publishing](https://github.com/mobilusoss/react-web-notification/issues/2)

### Ver 0.1.0 Initial release


test/components/Notification_spec.js
import React from 'react';
import ReactDom from 'react-dom';
import ReactTestUtils from 'react-dom/test-utils';

import chai from 'chai';
import sinon from 'sinon';
let expect = chai.expect;
import events from 'events'
let EventEmitter = events.EventEmitter;
import Notification from '../../src/components/Notification';

const PERMISSION_GRANTED = 'granted';
const PERMISSION_DENIED = 'denied';
const PERMISSION_DEFAULT = 'default';

describe('Test of Notification', () => {

  let component;

  beforeEach(() => {
  });

  describe('Notification component', () => {
    it('should have default properties', function () {
      component = ReactTestUtils.renderIntoDocument(<Notification title='test'/>);
      expect(component.props.ignore).to.be.eql(false);
      expect(component.props.disableActiveWindow).to.be.eql(false);
      expect(component.props.askAgain).to.be.eql(false);
      expect(typeof component.props.notSupported).to.be.eql('function');
      expect(typeof component.props.onPermissionGranted).to.be.eql('function');
      expect(typeof component.props.onPermissionDenied).to.be.eql('function');
      expect(typeof component.props.onShow).to.be.eql('function');
      expect(typeof component.props.onClick).to.be.eql('function');
      expect(typeof component.props.onClose).to.be.eql('function');
      expect(typeof component.props.onError).to.be.eql('function');
      expect(component.props.timeout).to.be.eql(5000);
      expect(component.props.options).to.be.empty;
    });

    it('should render dummy hidden tag', function () {
      component = ReactTestUtils.renderIntoDocument(<Notification title='test'/>);
      const el = ReactTestUtils.scryRenderedDOMComponentsWithTag(component, 'input');
      expect(el.length).to.be.eql(1);
      expect(ReactDom.findDOMNode(el[0]).type).to.be.eql('hidden');
    });
  });

  describe('Handling HTML5 Web Notification API', () => {
    describe('When Notification is not supported', () => {
      let cached, stub;

      before(() => {
        stub = sinon.stub(window.Notification, 'requestPermission');
        cached = window.Notification;
        window.Notification = null;
      });

      after(() => {
        window.Notification = cached;
        stub.restore();
      });

      it('should call notSupported prop', () => {
        let spy = sinon.spy();
        component = ReactTestUtils.renderIntoDocument(<Notification title='test' notSupported={spy}/>);
        expect(spy.calledOnce).to.be.eql(true);
        expect(stub.called).to.be.eql(false);
      });
    });

    describe('When Notification is supported', () => {
      describe('start request permission ', () =>{
        describe('When Notification is denied', () => {
          describe('Check callbacks', ()=> {

            let stub, spy1, spy2;
            before(() => {
              spy1 = sinon.spy();
              spy2 = sinon.spy();
              stub = sinon.stub(window.Notification, 'requestPermission').callsFake(function(cb){
                return cb(PERMISSION_DENIED);
              });
              component = ReactTestUtils.renderIntoDocument(<Notification title='test' notSupported={spy1} onPermissionDenied={spy2}/>);
            });

            after(() => {
              stub.restore();
            });

            it('should call window.Notification.requestPermission', () => {
              expect(stub.calledOnce).to.be.eql(true);
            });

            it('should call onPermissionDenied prop', () => {
              expect(spy1.called).to.be.eql(false);
              expect(spy2.calledOnce).to.be.eql(true);
            });
          });
        });

        describe('When Notification is already denied', () => {
          describe('Check callbacks', ()=> {

            let stub1, stub2, spy1, spy2;
            before(() => {
              spy1 = sinon.spy();
              spy2 = sinon.spy();
              stub1 = sinon.stub(window.Notification, 'permission').get(function getterFn() {
                return PERMISSION_DENIED;
              });
              stub2 = sinon.stub(window.Notification, 'requestPermission').callsFake(function(cb){
                return cb(PERMISSION_DENIED);
              });
              component = ReactTestUtils.renderIntoDocument(<Notification title='test' notSupported={spy1} onPermissionDenied={spy2}/>);
            });

            after(() => {
              stub1.restore();
              stub2.restore();
            });

            it('should not call window.Notification.requestPermission', () => {
              expect(stub2.calledOnce).to.be.eql(false);
            });

            it('should call onPermissionDenied prop', () => {
              expect(spy1.called).to.be.eql(false);
              expect(spy2.calledOnce).to.be.eql(true);
            });
          });
        });

        describe('When Notification is already denied, but `askAgain` prop is true', () => {
          describe('Check callbacks', ()=> {

            let stub1, stub2, spy1, spy2, spy3;
            before(() => {
              spy1 = sinon.spy();
              spy2 = sinon.spy();
              spy3 = sinon.spy();
              stub1 = sinon.stub(window.Notification, 'permission').get(function getterFn() {
                return PERMISSION_DENIED;
              });
              stub2 = sinon.stub(window.Notification, 'requestPermission').callsFake(function(cb){
                return cb(PERMISSION_GRANTED);
              });
              component = ReactTestUtils.renderIntoDocument(<Notification title='test' notSupported={spy1} onPermissionDenied={spy2} onPermissionGranted={spy3} askAgain={true}/>);
            });

            after(() => {
              stub1.restore();
              stub2.restore();
            });

            it('should call window.Notification.requestPermission', () => {
              expect(stub2.calledOnce).to.be.eql(true);
            });

            it('should call onPermissionGranted prop', () => {
              expect(spy1.called).to.be.eql(false);
              expect(spy2.called).to.be.eql(false);
              expect(spy3.calledOnce).to.be.eql(true);
            });
          });
        });

        describe('When Notification is granted', () => {
          let stub;
          before(() => {
            stub = sinon.stub(window.Notification, 'requestPermission').callsFake(function(cb){
              return cb(PERMISSION_GRANTED);
            });
          });

          after(() => {
            stub.restore();
          });

          describe('Check callbacks', ()=> {
            let spy1, spy2, spy3;
            before(() => {
              spy1 = sinon.spy();
              spy2 = sinon.spy();
              spy3 = sinon.spy();
              component = ReactTestUtils.renderIntoDocument(<Notification title='test' notSupported={spy1} onPermissionDenied={spy2} onPermissionGranted={spy3}/>);
            });

            it('should call window.Notification.requestPermission', () => {
              expect(stub.called).to.be.eql(true);
            });

            it('should call onPermissionDenied prop', () => {
              expect(spy1.called).to.be.eql(false);
              expect(spy2.called).to.be.eql(false);
              expect(spy3.calledOnce).to.be.eql(true);
            });
          });

          describe('Handle component properties', () => {

            let stubConstructor, onShowSpy, onClickSpy, onCloseSpy, onErrorSpy, ee;

            beforeEach(() => {
              EventEmitter.prototype.addEventListener = EventEmitter.prototype.addListener;
              ee = new EventEmitter();

              stubConstructor = sinon.stub(window, 'Notification');
              stubConstructor.onFirstCall().returns(ee);
            });

            afterEach(()=>{
              // stub.restore();
              stubConstructor.restore();
            });

            describe('when ignore prop is true', () => {
              onShowSpy = sinon.spy();
              onClickSpy = sinon.spy();
              onCloseSpy = sinon.spy();
              onErrorSpy = sinon.spy();

              it('does not trigger Notification', () => {
                component = ReactTestUtils.renderIntoDocument(<Notification title='test' ignore={true} onShow={onShowSpy} onClick={onClickSpy} onClose={onCloseSpy} onError={onErrorSpy} />);
                expect(stubConstructor.calledWithNew()).to.be.eql(false);
                expect(onShowSpy.called).to.be.eql(false);
                expect(onClickSpy.called).to.be.eql(false);
                expect(onCloseSpy.called).to.be.eql(false);
                expect(onErrorSpy.called).to.be.eql(false);
              });
            });

            describe('when ignore prop is false', () => {
              const MY_TITLE = 'mytitle';
              const MY_OPTIONS = {
                tag: 'mytag',
                body: 'mybody',
                icon: 'myicon',
                lang: 'en',
                dir: 'ltr'
              };
              onShowSpy = sinon.spy();
              onClickSpy = sinon.spy();
              onCloseSpy = sinon.spy();
              onErrorSpy = sinon.spy();

              it('trigger Notification with specified title and options', () => {
                component = ReactTestUtils.renderIntoDocument(<Notification title={MY_TITLE} options={MY_OPTIONS} ignore={false} onShow={onShowSpy} onClick={onClickSpy} onClose={onCloseSpy} onError={onErrorSpy} />);
                expect(stubConstructor.calledWithNew()).to.be.eql(true);
                expect(stubConstructor.calledWith(MY_TITLE, MY_OPTIONS)).to.be.eql(true);
              });

              it('call onShow prop when notification is shown', () => {
                let n = component._getNotificationInstance('mytag');
                n.onshow('showEvent');
                expect(onShowSpy.calledOnce).to.be.eql(true);
                let args = onShowSpy.args[0];
                expect(args[0]).to.be.eql('showEvent');
                expect(args[1]).to.be.eql('mytag');
                expect(onClickSpy.called).to.be.eql(false);
                expect(onCloseSpy.called).to.be.eql(false);
                expect(onErrorSpy.called).to.be.eql(false);
              });

              it('call onClick prop when notification is clicked', () => {
                let n = component._getNotificationInstance('mytag');
                n.onclick('clickEvent');
                expect(onClickSpy.calledOnce).to.be.eql(true);
                let args = onClickSpy.args[0];
                expect(args[0]).to.be.eql('clickEvent');
                expect(args[1]).to.be.eql('mytag');
              });

              it('call onCLose prop when notification is closed', () => {
                let n = component._getNotificationInstance('mytag');
                n.onclose('closeEvent');
                expect(onCloseSpy.calledOnce).to.be.eql(true);
                let args = onCloseSpy.args[0];
                expect(args[0]).to.be.eql('closeEvent');
                expect(args[1]).to.be.eql('mytag');
              });

              it('call onError prop when notification throw error', () => {
                let n = component._getNotificationInstance('mytag');
                n.onerror('errorEvent');
                expect(onErrorSpy.calledOnce).to.be.eql(true);
                let args = onErrorSpy.args[0];
                expect(args[0]).to.be.eql('errorEvent');
                expect(args[1]).to.be.eql('mytag');
              });
            });
            describe('test of autoClose timer', () => {
              const MY_TITLE = 'mytitle';
              const MY_OPTIONS = {
                tag: 'mytag',
                body: 'mybody',
                icon: 'myicon',
                lang: 'en',
                dir: 'ltr'
              };
              describe('when `props.timeout` is less than eql 0', () => {
                let n;
                before(() => {
                  component = ReactTestUtils.renderIntoDocument(<Notification title={MY_TITLE} options={MY_OPTIONS} ignore={false} timeout={0}/>);
                  n = component._getNotificationInstance('mytag');
                  sinon.stub(n, 'close');
                  n.onshow('showEvent');
                });
                after(() => {
                  n.close.restore();
                })
                it('will not trigger close automatically', (done) => {
                  setTimeout(() => {
                    expect(n.close.called).to.be.eql(false);
                    done();
                  }, 200);
                })
              })
              describe('when `props.timeout` is greater than 0', () => {
                let n;
                before(() => {
                  component = ReactTestUtils.renderIntoDocument(<Notification title={MY_TITLE} options={MY_OPTIONS} ignore={false} timeout={50}/>);
                  n = component._getNotificationInstance('mytag');
                  sinon.stub(n, 'close');
                  n.onshow('showEvent');
                });
                after(() => {
                  n.close.restore();
                });
                it('will trigger close automatically', (done) => {
                  setTimeout(() => {
                    expect(n.close.called).to.be.eql(true);
                    done();
                  }, 200);
                })
              })
            })
            describe('when swRegistration prop is defined', () => {
              const swRegistrationMock = { showNotification: sinon.stub().resolves({ notification: ee }) }
              const MY_TITLE = 'mytitle';
              const MY_OPTIONS = {
                tag: 'mytag',
                body: 'mybody',
                icon: 'myicon',
                lang: 'en',
                dir: 'ltr',
                requireInteraction: true,
              };

              it('does not trigger Notification but trigger swRegistration.showNotification', () => {
                component = ReactTestUtils.renderIntoDocument(<Notification title={MY_TITLE} options={MY_OPTIONS} swRegistration={swRegistrationMock} />);
                expect(stubConstructor.calledWithNew()).to.be.eql(false);
                expect(swRegistrationMock.showNotification.calledWith(MY_TITLE, MY_OPTIONS)).to.be.eql(true);
              });
            });
          });
        });
      });
    });
  });
});


karma.conf.js
// Karma configuration
// Generated on Fri May 15 2015 12:31:20 GMT+0900 (JST)

module.exports = function(config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',


    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['browserify','mocha'],


    // list of files / patterns to load in the browser
    files: [
      'test/**/*_spec.js',
      { pattern: 'node_modules/sinon/pkg/sinon.js', watched: false, included: true }
    ],


    // list of files to exclude
    exclude: [
      'test/_helper/*.js'
    ],


    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
      'test/**/*_spec.js': ['browserify'],
      'src/**/*.js': ['coverage']
    },

    browserify: {
      debug: true,
      'transform': [
        [
          'babelify',{
            "presets": [
              ["@babel/env",{
                  "useBuiltIns": "entry",
                  "corejs": 3
                }
              ],
              "@babel/react"
            ],
            "plugins": [
              "istanbul",
            ],
          }
        ]
      ]
    },

    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['spec', 'coverage'],

     coverageReporter: {
      reporters: [
        // reporters not supporting the `file` property
        {type:'lcovonly', subdir: '.'},
        { type: 'text' },
      ]
    },

    // client: {
    //   mocha: {
    //     // reporter: 'html', // change Karma's debug.html to the mocha web reporter
    //     // ui: 'tdd'
    //   }
    // },

    // web server port
    port: 9876,


    // enable / disable colors in the output (reporters and logs)
    colors: true,


    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,


    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,


    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    // browsers: ['Chrome', 'Firefox', 'PhantomJS', 'IE'],
    browsers: [process.env.CONTINUOUS_INTEGRATION ? 'Firefox' : 'Chrome'],


    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: false
  });
};


example/app.js
'use strict';

import React from 'react';
import ReactDom from 'react-dom';
import Notification  from '../lib/components/Notification';

//allow react dev tools work
window.React = React;

class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      ignore: true,
      title: ''
    };
  }

  handlePermissionGranted(){
    console.log('Permission Granted');
    this.setState({
      ignore: false
    });
  }
  handlePermissionDenied(){
    console.log('Permission Denied');
    this.setState({
      ignore: true
    });
  }
  handleNotSupported(){
    console.log('Web Notification not Supported');
    this.setState({
      ignore: true
    });
  }

  handleNotificationOnClick(e, tag){
    console.log(e, 'Notification clicked tag:' + tag);
  }

  handleNotificationOnError(e, tag){
    console.log(e, 'Notification error tag:' + tag);
  }

  handleNotificationOnClose(e, tag){
    console.log(e, 'Notification closed tag:' + tag);
  }

  handleNotificationOnShow(e, tag){
    this.playSound();
    console.log(e, 'Notification shown tag:' + tag);
  }

  playSound(filename){
    document.getElementById('sound').play();
  }

  handleButtonClick() {

    if(this.state.ignore) {
      return;
    }

    const now = Date.now();

    const title = 'React-Web-Notification' + now;
    const body = 'Hello' + new Date();
    const tag = now;
    const icon = 'http://mobilusoss.github.io/react-web-notification/example/Notifications_button_24.png';
    // const icon = 'http://localhost:3000/Notifications_button_24.png';

    // Available options
    // See https://developer.mozilla.org/en-US/docs/Web/API/Notification/Notification
    const options = {
      tag: tag,
      body: body,
      icon: icon,
      lang: 'en',
      dir: 'ltr',
      sound: './sound.mp3'  // no browsers supported https://developer.mozilla.org/en/docs/Web/API/notification/sound#Browser_compatibility
    }
    this.setState({
      title: title,
      options: options
    });
  }

  handleButtonClick2() {
    this.props.swRegistration.getNotifications({}).then(function(notifications) {
      console.log(notifications);
    });
  }

  render() {
    return (
      <div>
        <button onClick={this.handleButtonClick.bind(this)}>Notif!</button>
        {document.title === 'swExample' && <button onClick={this.handleButtonClick2.bind(this)}>swRegistration.getNotifications</button>}
        <Notification
          ignore={this.state.ignore && this.state.title !== ''}
          notSupported={this.handleNotSupported.bind(this)}
          onPermissionGranted={this.handlePermissionGranted.bind(this)}
          onPermissionDenied={this.handlePermissionDenied.bind(this)}
          onShow={this.handleNotificationOnShow.bind(this)}
          onClick={this.handleNotificationOnClick.bind(this)}
          onClose={this.handleNotificationOnClose.bind(this)}
          onError={this.handleNotificationOnError.bind(this)}
          timeout={5000}
          title={this.state.title}
          options={this.state.options}
          swRegistration={this.props.swRegistration}
        />
        <audio id='sound' preload='auto'>
          <source src='./sound.mp3' type='audio/mpeg' />
          <source src='./sound.ogg' type='audio/ogg' />
          <embed hidden={true} autostart='false' loop={false} src='./sound.mp3' />
        </audio>
      </div>
    )
  }
};
if (document.title === 'swExample') {
  navigator.serviceWorker.register('sw.js')
  .then(function(registration) {
    ReactDom.render(<App swRegistration={registration}/>, document.getElementById('out'));
  });
} else {
  ReactDom.render(<App/>, document.getElementById('out'));
}



src/components/Notification.js
import React from 'react';
import { bool, func, number, object, string } from 'prop-types';

const PERMISSION_GRANTED = 'granted';
const PERMISSION_DENIED = 'denied';

const seqGen = () => {
  let i = 0;
  return () => {
    return i++;
  };
};
const seq = seqGen();

// https://developer.mozilla.org/en-US/docs/Web/API/Notifications_API/Using_the_Notifications_API
// https://github.com/mobilusoss/react-web-notification/issues/66
function checkNotificationPromise() {
  try {
    window.Notification.requestPermission().then();
  } catch(e) {
    return false;
  }
  return true;
}
class Notification extends React.Component {
  constructor(props) {
    super(props);

    let supported = false;
    let granted = false;
    if (('Notification' in window) && window.Notification) {
      supported = true;
      if (window.Notification.permission === PERMISSION_GRANTED) {
        granted = true;
      }
    }

    this.state = {
      supported: supported,
      granted: granted
    };
    // Do not save Notification instance in state
    this.notifications = {};
    this.windowFocus = true;
    this.onWindowFocus = this._onWindowFocus.bind(this);
    this.onWindowBlur = this._onWindowBlur.bind(this);
  }

  _onWindowFocus(){
    this.windowFocus = true;
  }

  _onWindowBlur(){
    this.windowFocus = false;
  }

  _askPermission(){
    const handlePermission = (permission) => {
      let result = permission === PERMISSION_GRANTED;
      this.setState({
        granted: result
      }, () => {
        if (result) {
          this.props.onPermissionGranted();
        } else {
          this.props.onPermissionDenied();
        }
      });
    }
    if (checkNotificationPromise()) {
      window.Notification.requestPermission()
      .then((permission) => {
        handlePermission(permission);
      })
    } else {
      window.Notification.requestPermission((permission) => {
        handlePermission(permission);
      });
    }
  }

  componentDidMount(){
    if (this.props.disableActiveWindow) {
      window.addEventListener('focus', this.onWindowFocus);
      window.addEventListener('blur', this.onWindowBlur);
    }

    if (!this.state.supported) {
      this.props.notSupported();
    } else if (this.state.granted) {
      this.props.onPermissionGranted();
    } else {
      if (window.Notification.permission === PERMISSION_DENIED){
        if (this.props.askAgain){
          this._askPermission();
        } else {
          this.props.onPermissionDenied();
        }
      } else {
        this._askPermission();
      }
    }
  }

  componentWillUnmount(){
    if (this.props.disableActiveWindow) {
      window.removeEventListener('focus', this.onWindowFocus);
      window.removeEventListener('blur', this.onWindowBlur);
    }
  }

  doNotification() {
    let opt = this.props.options;
    if (typeof opt.tag !== 'string') {
      opt.tag = 'web-notification-' + seq();
    }
    if (this.notifications[opt.tag]) {
      return;
    }

    if (this.props.swRegistration && this.props.swRegistration.showNotification) {
      this.props.swRegistration.showNotification(this.props.title, opt)
      this.notifications[opt.tag] = {};
    } else {
      const n = new window.Notification(this.props.title, opt);
      n.onshow = e => {
        this.props.onShow(e, opt.tag);
        if (this.props.timeout > 0) {
          setTimeout(() => {
            this.close(n);
          }, this.props.timeout);
        }
      };
      n.onclick = e => { this.props.onClick(e, opt.tag); };
      n.onclose = e => { this.props.onClose(e, opt.tag); };
      n.onerror = e => { this.props.onError(e, opt.tag); };

      this.notifications[opt.tag] = n;
    }
  }

  render() {
    let doNotShowOnActiveWindow = this.props.disableActiveWindow && this.windowFocus;
    if (!this.props.ignore && this.props.title && this.state.supported && this.state.granted && !doNotShowOnActiveWindow) {
      this.doNotification();
    }

    // return null cause
    // Error: Invariant Violation: Notification.render(): A valid ReactComponent must be returned. You may have returned undefined, an array or some other invalid object.
    return (
      <input type='hidden' name='dummy-for-react-web-notification' style={{display: 'none'}} />
    );
  }

  close(n) {
    if (n && typeof n.close === 'function') {
      n.close();
    }
  }

  // for debug
  _getNotificationInstance(tag) {
    return this.notifications[tag];
  }
}

Notification.propTypes = {
  ignore: bool,
  disableActiveWindow: bool,
  askAgain: bool,
  notSupported: func,
  onPermissionGranted: func,
  onPermissionDenied: func,
  onShow: func,
  onClick: func,
  onClose: func,
  onError: func,
  timeout: number,
  title: string.isRequired,
  options: object,
  swRegistration: object,
};

Notification.defaultProps = {
  ignore: false,
  disableActiveWindow: false,
  askAgain: false,
  notSupported: () => {},
  onPermissionGranted: () => {},
  onPermissionDenied: () => {},
  onShow: () => {},
  onClick: () => {},
  onClose: () => {},
  onError: () => {},
  timeout: 5000,
  options: {},
  swRegistration: null,
};

export default Notification;
